src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java

```diff
@@ -0,0 +1,568 @@
+          if (result == Result.COMPLETE) {
+            isReadingChunk = false;
+          } else if (result == Result.ERROR) {
+            throw new PatchFailedException(
+                "Wrong chunk detected near line " + (i + 1) + ": " + line +
+                    ", does not expect a deleting line here.");
+          }
+          break;
+        case CHUNK_EQL:
+          oldLineCount++;
+          newLineCount++;
+          patchContent.add(line);
+          result = header.check(oldLineCount, newLineCount);
+          if (result == Result.COMPLETE) {
+            isReadingChunk = false;
+          } else if (result == Result.ERROR) {
+            throw new PatchFailedException(
+                "Wrong chunk detected near line " + (i + 1) + ": " + line +
+                    ", does not expect an equal line here.");
+          }
+          break;
+        case RENAME_FROM:
+          hasRenameFrom = true;
+          if (oldFileStr == null) {
+            // len("rename from ") == 12
+            oldFileStr = line.substring(12).trim();
+            if (oldFileStr.isEmpty()) {
+              throw new PatchFailedException(
+                  String.format("Cannot determine file name from line %d:\n%s", i + 1, line));
+            }
+            oldFile = getFilePath(oldFileStr, outputDirectory, i + 1);
+          }
+          break;
+        case RENAME_TO:
+          hasRenameTo = true;
+          if (newFileStr == null) {
+            // len("rename to ") == 10
+            newFileStr = line.substring(10).trim();
+            if (newFileStr.isEmpty()) {
+              throw new PatchFailedException(
+                  String.format("Cannot determine file name from line %d:\n%s", i + 1, line));
+            }
+            newFile = getFilePath(newFileStr, outputDirectory, i + 1);
+          }
+          break;
+        case GIT_LINE:
+          break;
+        case GIT_HEADER:
+        case UNKNOWN:
+          // A git header line or an unknown line should trigger an action to apply collected
+          // patch content to a file.
+
+          // Renaming is a git only format
+          boolean isRenaming = isGitDiff && hasRenameFrom && hasRenameTo;
+
+          if (!patchContent.isEmpty() || isRenaming) {
+            // We collected something useful, let's do some sanity checks before applying the patch.
+            checkPatchContentIsComplete(patchContent, header, oldLineCount, newLineCount, i + 1);
+
+            int patchStartLocation = i + 1 - patchContent.size();
+
+            if (isRenaming) {
+              checkFilesStatusForRenaming(
+                  oldFile, newFile, oldFileStr, newFileStr, patchStartLocation);
+            }
+
+            Patch<String> patch = DiffUtils.parseUnifiedDiff(patchContent);
+            checkFilesStatusForPatching(
+                patch, oldFile, newFile, oldFileStr, newFileStr, patchStartLocation);
+
+            applyPatchToFile(patch, oldFile, newFile, isRenaming);
+          }
+
```

Comment:
More check

Reply:
It turned out even we only saw `+++ ` and `--- ` lines without `@@ -l,s +l,s @@` line, `patchContent` won't be empty, because both lines will be added into `patchContent`. 
But I added the check and test anyway, please take a look! ;)